Add some clarity to contributors file (#710)
ClassicPress/ClassicPress: .github/CONTRIBUTING.md

Before contributing to this repository, please read our [Democracy page](https://www.classicpress.net/democracy/) to understand how we make decisions about what is included in ClassicPress, and our [roadmap](https://www.classicpress.net/roadmap/) to see what's already planned for the next couple of releases.

The **petitions process** mentioned in our Democracy document is very important to help us prioritize new features based on the needs of our users. There are some [exceptions](https://www.classicpress.net/democracy/#democracy-exceptions) to this process for minor changes and bugfixes, but generally speaking it is a good idea to search or ask in one of our communication channels (see below) before undertaking a change, because most changes should go through the petitions process.

Also, please be sure to follow our [code of conduct](https://www.classicpress.net/democracy/#democracy-conduct) in all interactions with ClassicPress community members.

> ## Quick Links

Reviewer: ```suggestion
## Table of Contents

```

"Quick Links" makes it sound like this goes out to another part of the document.

Author: Debated it too. Committed. Thanks.

> - [Contributions](#contributions)
>   - [Quick Links](#quick-links)
>   - [Communication channels](#communication-channels)
>   - [Review criteria](#review-criteria)
>   - [What to work on?](#what-to-work-on)
>   - [Setting up Local Testing and Dev Environment](#setting-up-local-testing-and-dev-environment)
>   - [More information](#more-information)
>     - [Tips for good Pull Requests (PRs)](#tips-for-good-pull-requests-prs)
>     - [Automated tests](#automated-tests)
>     - [Backporting changes from WordPress](#backporting-changes-from-wordpress)
>     - [Tips for a good Backport PR.](#tips-for-a-good-backport-pr)
>     - [Making a Backport PR.](#making-a-backport-pr)
>     - [How to review a PR](#how-to-review-a-pr)

Reviewer: No need to link to the main heading or the table of contents itself.

```suggestion
- [Communication channels](#communication-channels)
- [Review criteria](#review-criteria)
- [What to work on?](#what-to-work-on)
- [Setting up Local Testing and Dev Environment](#setting-up-local-testing-and-dev-environment)
- [More information](#more-information)
  - [Tips for good Pull Requests (PRs)](#tips-for-good-pull-requests-prs)
  - [Automated tests](#automated-tests)
  - [Backporting changes from WordPress](#backporting-changes-from-wordpress)
  - [Tips for a good Backport PR.](#tips-for-a-good-backport-pr)
  - [Making a Backport PR.](#making-a-backport-pr)
  - [How to review a PR](#how-to-review-a-pr)
```

  1. The change impacts existing ClassicPress users. Otherwise, there are literally thousands of things we could look at, but we need to prioritize our development time. Right now the best tool we have for this is [petitions](https://forums.classicpress.net/c/governance/petitions/77).
  2. The change is not going to break backward compatibility, and has minimal effects on the existing plugin and theme ecosystem. A good way to evaluate the effects of a change on plugins or themes is to do a search on [wpdirectory](https://wpdirectory.net). (Major changes are also a possibility but require a planning effort around when and how they will be released, as well as agreement from the community per our [democratic process](https://www.classicpress.net/democracy/).)
  3. The change has automated tests.
  4. We understand the code change very well or can ask questions of someone who understands it very well.

If your change meets all of these criteria then we will most likely accept it.

## What to work on?

If you're not sure where to start contributing, here are some ideas:

>   - Review and test [existing PRs](https://github.com/ClassicPress/ClassicPress/pulls), especially looking at how well they fit the criteria described above. Let us know how you tested the changes and what you found. Screenshots, gifs, video and text instructions documenting the tests can be very useful. We need to be thorough and careful with any changes we make, and the more eyes we can get on our PRs the better.

Reviewer: ```suggestion
  - Review and test [existing PRs](https://github.com/ClassicPress/ClassicPress/pulls), especially looking at how well they fit the criteria described above. Let us know how you tested the changes and what you found. We need to be thorough and careful with any changes we make, and the more eyes we can get on our PRs the better. Screenshots and text instructions documenting the status and testing of all PRs are very useful, as well as videos and gifs if applicable. These help us know the status and completion of each PR so that we can be thorough and careful with any changes we make, and the more eyes we can get on our PRs the better.
```

# Contributions

ClassicPress is a volunteer-driven open source project that thrives when we have multiple contributors working towards a common goal. 

> There may be some delays in reviewing and merging pull requests. This is because we take a lot of care with our platform and ClassicPress is maintained by volunteers in our free time.

Reviewer: How about just removing this sentence instead? The delay for reviewing PRs should be shortened when we have more people looking at them

Author: 50/50 for me. One of the most unexpected things in FOSS is people assuming everything is snap. This is a gentle reminder. But we can agree to disagree.

Reviewer: Maybe we can make the reminder more gentle then:

```suggestion
We invite all contributors to not only submit pull requests that are in line with the goals and direction of the project, but also to discuss and review other contributors' issues and pull requests. This will help us make the most of our limited time and review all contributions quickly.
```

## What to work on?

If you're not sure where to start contributing, here are some ideas:

  - Review and test [existing PRs](https://github.com/ClassicPress/ClassicPress/pulls), especially looking at how well they fit the criteria described above. Let us know how you tested the changes and what you found. We need to be thorough and careful with any changes we make, and the more eyes we can get on our PRs the better.
  - Review and test [existing PRs](https://github.com/ClassicPress/ClassicPress/pulls), especially looking at how well they fit the criteria described above. Let us know how you tested the changes and what you found. Screenshots, gifs, video and text instructions documenting the tests can be very useful. We need to be thorough and careful with any changes we make, and the more eyes we can get on our PRs the better.
  - Bugfixes, you can find a list of known bugs on our [issues page](https://github.com/ClassicPress/ClassicPress/issues).
  - Issues with the [`help wanted`](https://github.com/ClassicPress/ClassicPress/labels/help%20wanted) or [`good first issue`](https://github.com/ClassicPress/ClassicPress/labels/good%20first%20issue) labels.

>   - [Planned](https://petitions.classicpress.net/?view=planned) or [most wanted](https://petitions.classicpress.net/?view=most-wanted) feature requests from our petitions site.

Reviewer: Why was this part removed?

Author: We have since moved the petitions to forums. We haven't yet found a way to show the most wanted and planned. Milestones on Github might be the next best thing to share this until we figure the forums route out.

Author: Let's keep the conversation to this via https://github.com/ClassicPress/ClassicPress/pull/710#discussion_r619798985

>   - [Planned Milestones](https://github.com/ClassicPress/ClassicPress/milestones) or exploratory PRs with your own suggested changes. Please remember these will be subject to review to make sure they are in line with the project's direction.

Reviewer: I think adding milestones is good but it should probably go with the sentence above instead (the one that was removed).

Author: I guess we can do that for now until we retire the petitions subdomain. However, it now says, "it is archived." All the petitions are on the forums. We are using "hearts" to poll until a properway to do this is chosen. Would you still suggest we keep pointing to this?

Reviewer: OK, I see why that was removed. The archive domain will stay up (to keep old links working, point people to the new location, and show a better view of what people decided they wanted to see in ClassicPress), but it will become more outdated over time. Also, jumping right in with those petitions is probably not the best place to get started with contributing.

Author: I think that resolves it.

Changes must be proposed and reviewed **individually** - long lists of tickets or changesets from individual WordPress versions don't give us what we need in order to plan and understand each change well.

If you're not sure about any of that, then it's a good idea to ask first. A good way is to create an issue for the specific change you're interested in, along with links to the relevant WordPress changesets and tickets, and any other information you have about how the change works.

There are some changes that we already know we want to backport because they fit into our plans for future versions of ClassicPress. You can find some examples under the [`WP backport` label](https://github.com/ClassicPress/ClassicPress/labels/WP%20backport).

### Tips for a good Backport PR.
- Give the PRs a clear title. Explain in a pull request or issue what the specific change is, how it works and why ClassicPress in particular should adopt it. Often there is a lot of good discussion about each change in the relevant WordPress Trac tickets. It makes the maintainers' job much easier to see this discussion summarized and linked.

> - Explain what testing has been done on the PR and what may be left to do. Screenshots, gif, video and text instructions documenting the tests can be very useful. Those tell other contributors looking at the PR how to verify the changes.

Reviewer: ```suggestion
- Explain what testing has been done on the PR and what may be left to do. Screenshots, gif, video and text instructions documenting the tests can be very useful. Those tell other contributors looking at the PR how to verify the changes. "This was tested in WordPress" **is not enough** to meet our review criteria, see [Review Criteria](#review-criteria) above.
```

- Look to make sure the backport is done correctly, with attribution to the original WP committer. 
- If there are merge conflicts during the backport, check if they have been resolved correctly by comparing the final changes from the full PR against the corresponding WP changeset(s). Remember to fix these **in a separate commit**. If you cannot resolve the conflicts, ask for help in the [Core slack channel](https://www.classicpress.net/join-slack/) or just push your branch as-is and we'll take care of it!

### Making a Backport PR.

> When you're ready to backport a code change, make a new branch to your ClassicPress fork:

Reviewer 2: > make a new branch to your ClassicPress fork:

This sounds like the next thing will show how to make a branch in the fork, but the next thing is about identifying a changeset (sort of).  Does the script make a branch? If so, say so. If not, put the command to make the branch.

Author: https://github.com/ClassicPress/ClassicPress/pull/710#discussion_r619825795

- Identify the WordPress **changeset number** that you'd like to port such as `43123`.
- Run `bin/backport-wp-commit.sh` script in your terminal/command prompt.

```
bin/backport-wp-commit.sh "changeset-number"

> ```

Reviewer 2: Are the quotes needed? Can you put the actual number in the command?

Reviewer: The changeset number is the number of the WP changeset to backport. The user of the script will put the actual number in the command.

I usually use the format `CHANGESET_NUMBER` for this. Quotes are probably best avoided since they have special meaning on the command line.

Author: Made changes and used a block letter variable.

 to apply the change to the code. 
 
 If you're porting multiple changesets, you can use the `-c` option to this script 

```

> bin/backport-wp-commit.sh -c "changeset-number"

Reviewer 2: This example is for multiple changesets, so please show the correct syntax.

Reviewer: The syntax is correct (changesets must be backported one at a time), but it can be explained better that to backport multiple changesets you do the following:

1. Start a new branch
2. `bin/backport-wp-commit.sh -c CHANGESET_NUMBER`
3. Resolve merge conflicts (if any)
4. Repeat steps 2 and 3

Author: Made some changes

```
to apply all the changesets to the same branch. 

> Then push your branch for review as a PR.

Reviewer 2: Say how to push your branch. At least mention it's your fork, not the ClassicPress repo.

Author: Agree this might have to change with the new section of making your local dev environment. Let's fix them hand in hand.

> Using this script for all backports saves time for you and for the maintainers. As it uses a standardized format for commit messages making it possible for us to track which WordPress changes we've already included.

Reviewer: Fine to put this in two separate sentences but grammar is a bit off.

```suggestion
Using this script for all backports saves time for you and for the maintainers. It uses a standardized format for commit messages, which makes it possible for us to track which WordPress changes we've already included.
```

> **Note:** If there are merge conflicts for your changes, the script will let you know. Please fix the conflicts **in a separate commit** so that we can take a look at the conflicts and their fixes, or if you're not comßfortable resolving merge conflicts, just push your branch as-is and we'll take care of it!

Reviewer: Extra non-English letter inserted here:

```suggestion
**Note:** If there are merge conflicts for your changes, the script will let you know. Please fix the conflicts **in a separate commit** so that we can take a look at the conflicts and their fixes, or if you're not comfortable resolving merge conflicts, just push your branch as-is and we'll take care of it!
```

Changes must be proposed and reviewed **individually** - long lists of tickets or changesets from individual WordPress versions don't give us what we need in order to plan and understand each change well.

If you're not sure about any of that, then it's a good idea to ask first. A good way is to create an issue for the specific change you're interested in, along with links to the relevant WordPress changesets and tickets, and any other information you have about how the change works.

There are some changes that we already know we want to backport because they fit into our plans for future versions of ClassicPress. You can find some examples under the [`WP backport` label](https://github.com/ClassicPress/ClassicPress/labels/WP%20backport).

### Tips for a good Backport PR.
- Give the PRs a clear title. Explain in a pull request or issue what the specific change is, how it works and why ClassicPress in particular should adopt it. Often there is a lot of good discussion about each change in the relevant WordPress Trac tickets. It makes the maintainers' job much easier to see this discussion summarized and linked.

> - Explain what testing has been done on the PR and what may be left to do. Screenshots, gif, video and text instructions documenting the tests can be very useful. Those tell other contributors looking at the PR how to verify the changes. "This was tested in WordPress" **is not enough** to meet our review criteria, see [Review Criteria](#review-criteria) above.

Reviewer: ```suggestion
- Explain what testing has been done on the PR and what may be left to do. Screenshots and text instructions documenting the changes and how to test them are very useful. Those tell other contributors looking at the PR how to verify the changes. "This was tested in WordPress" **is not enough** to meet our review criteria, see [Review Criteria](#review-criteria) above.
- Look through the WordPress tickets and commits associated with these changes. Make sure there are no other changes to commit, no outstanding issues with the WordPress changes mentioned in Trac tickets, and summarize your findings on your PR.
```

### Making a Backport PR.
When you're ready to backport a code change, make a new branch to your ClassicPress fork using:
```
git checkout upstream/develop -B branch-name
```

See [how to set up local environment](#setting-up-local-testing-and-dev-environment).

1. Identify the WordPress **changeset number** that you'd like to port such as `43123`.
2. Run `bin/backport-wp-commit.sh` script in your terminal/command prompt.

```
bin/backport-wp-commit.sh CHANGESET_NUMBER
```
 to apply the change to the code. 
 
 If you're porting multiple changesets, you can use the `-c` option to this script 

```
bin/backport-wp-commit.sh -c CHANGESET_NUMBER
```
to apply all the changesets to the same branch. 

3. Resolve merge conflicts (if any)
4. Repeat steps 2 and 3
5. Then push your branch to your fork online using:
   
```
git checkout upstream/develop -B desired-branch-name
```
6. Use the online platform to make a PR to the `develop` branch for review.

> !["Opening a PR to the develop branch"][open_pr_br_develop]

Reviewer: This is not the right image. Maybe let's leave that out for now?

Author: Will revert this. Should I squash changes?

Reviewer: No need to squash changes, they will get rolled up into one commit when we merge this.
